feat(datapath/hubble): IP Packet Tracing Implementation

sypakine opened the pull request:
Related: #35057

This PR includes the initial IP packet tracing implementation (#35057), including:

- parse IP option in datapath
- embed IP trace ID  in datapath event
- parse IP trace ID from datapath event into version-bumped `drop` and `trace` control plane structs
- expose IP trace ID in cilium monitor and hubble flow output

This PR addresses the backwards-compatibility issue (downgraded agent parsing new dataplane events) present in [PR 33957](https://github.com/cilium/cilium/pull/33957).

```release-note
Implement IP Tracing via IPv4 Option
```

asauber reviewed (approved):
The new flag looks good to me and is properly documented.

Bigdelle committed:
feat: add IPv4 packet tracing via IP options.

- Added feature-gate ip-tracing-option-type to enable and set IP option tracing.
- Implemented helper functions to parse IPv4 options.
- Added logic to save parsed IPv4 options into a per-CPU array map.

Signed-off-by: Ben Bigdelle <bigdelle@google.com>

sypakine committed:
feat: add IPTraceID into flow struct

- Modified flow.proto to include IPTraceID struct
- Add IPTraceID field to Flow and FlowFilter struct
- Add parsing to decode IPTraceID into Flow

Signed-off-by: Mark St John <markstjohn@google.com>

sypakine committed:
feat: expose IP trace in trace notify messages

This change:
- parses the trace ID from the trace notify dataplane struct and inserts in the control plane struct
- respsects the dataplane version and parses the data accordingly
- output trace ID in cilium monitor output

Signed-off-by: Mark St John <markstjohn@google.com>

Bigdelle committed:
feat: add IP option tracing parsing in ingress programs

- Implemented parsing logic for IP options in ingress programs.
- Store extracted IP options into per-CPU array map.

Signed-off-by: Ben Bigdelle <bigdelle@google.com>

sypakine committed:
test: update drop notify test to accommodate versioning

Separating the structural changes to the test from changes to the implementation.

Signed-off-by: Mark St John <markstjohn@google.com>

sypakine committed:
feat: expose IP trace in drop notify struct

This change:
- parses the trace ID from the drop notify dataplane struct and inserts in the control plane struct.
- respsects the dataplane version and parses the data accordingly
- output trace ID in cilium monitor output

Signed-off-by: Mark St John <markstjohn@google.com>

sypakine committed:
feat: add IP trace ID filter to hubble

- Add filter flag --ip-trace-id to hubble.
- Add IPTrace output to hubble printer.
- Update testing for observe help command.

Signed-off-by: Mark St John <markstjohn@google.com>
